src/main/java/com/teragrep/rlp_03/context/buffer/BufferLeaseImpl.java

```diff
@@ -1,127 +1,100 @@
+    private final Phaser phaser;
+
+    private final BufferLeasePool bufferLeasePool;
+
+    public BufferLeaseImpl(BufferContainer bc, BufferLeasePool bufferLeasePool) {
+        this.bufferContainer = bc;
+        this.bufferLeasePool = bufferLeasePool;
+
+        this.phaser = new ClearingPhaser(1); // registered = 1
+    }
+
+    @Override
+    public BufferContainer bufferContainer() {
+        return this.bufferContainer;
     }
 
     @Override
     public long id() {
-        return id;
+        return this.bufferContainer.id();
     }
 
     @Override
     public long refs() {
-        lock.lock();
-        try {
-            return refCount;
-        }
-        finally {
-            lock.unlock();
-        }
+        // initial number of registered parties is 1
+        return phaser.getRegisteredParties();
     }
 
     @Override
     public ByteBuffer buffer() {
-        lock.lock();
-        try {
-            return buffer;
-        }
-        finally {
-            lock.unlock();
-        }
-
+        return this.bufferContainer.buffer();
     }
 
     @Override
     public void addRef() {
-        lock.lock();
-        try {
-            refCount++;
-        }
-        finally {
-            lock.unlock();
+        int a = this.phaser.register();
+        if (a < 0) {
+            throw new IllegalStateException("Cannot add reference, BufferLease phaser was already terminated!");
         }
     }
 
     @Override
     public void removeRef() {
-        lock.lock();
-        try {
-
-            long newRefs = refCount - 1;
-            if (newRefs < 0) {
-                throw new IllegalStateException("refs must not be negative");
-            }
-
-            refCount = newRefs;
-        }
-        finally {
-            lock.unlock();
+        int a = phaser.arriveAndDeregister();
+        if (a < 0) {
+            throw new IllegalStateException("Cannot remove reference, BufferLease phaser was already terminated!");
         }
     }
 
     @Override
     public boolean isRefCountZero() {
-        lock.lock();
-        try {
-            return refCount == 0;
-        }
-        finally {
-            lock.unlock();
-        }
+        return phaser.isTerminated();
     }
 
-
     @Override
-    public String toString() {
-        lock.lock();
-        try {
-            return "BufferLease{" +
-                    "buffer=" + buffer +
-                    ", refCount=" + refCount +
-                    '}';
-        } finally {
-            lock.unlock();
-        }
+    public boolean isStub() {
+        return this.bufferContainer.isStub();
     }
 
     @Override
-    public boolean isStub() {
-        LOGGER.debug("id <{}>", id);
-        return false;
+    public synchronized boolean attemptRelease() {
+        /*
+        int a = phaser.arriveAndDeregister();
+        if (a < 0) {
+            throw new IllegalStateException("poks");
+        }
+        return phaser.isTerminated();
+
+         */
```

Comment:
Remove commented out code, doesn't look like something that should stay here for future reference

Reply:
fix [c447182](https://github.com/teragrep/rlp_03/pull/84/commits/c44718257663e780cb7e9699b9c1159fcf9517e9)